File: client-launcher/src/main/java/hivens/launcher/ServerListService.java
package hivens.launcher;

import hivens.core.api.IServerListService;
import hivens.core.api.SmartyNetworkService;
import hivens.core.api.dto.SmartyServer;
import hivens.core.api.model.ServerProfile;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.CompletableFuture;

public class ServerListService implements IServerListService {

    private final SmartyNetworkService networkService;

    // Кэш в памяти
    private List<ServerProfile> cachedProfiles = null;

    public ServerListService() {
        this.networkService = new SmartyNetworkService();
    }

    public CompletableFuture<List<ServerProfile>> fetchProfiles() {
        // Если кэш есть - отдаем мгновенно
        if (cachedProfiles != null && !cachedProfiles.isEmpty()) {
            return CompletableFuture.completedFuture(cachedProfiles);
        }

        return CompletableFuture.supplyAsync(() -> {
            List<ServerProfile> profiles = new ArrayList<>();
            try {
                List<SmartyServer> rawServers = networkService.getServers();
                for (SmartyServer srv : rawServers) {
                    profiles.add(getProfile(srv));
                }
                // Сохраняем в кэш
                this.cachedProfiles = profiles;
            } catch (Exception e) {
                System.err.println("Error fetching profiles: " + e.getMessage());
            }
            return profiles;
        });
    }

    // Метод для принудительного обновления (например, кнопка "Обновить")
    public void clearCache() {
        this.cachedProfiles = null;
    }

    private static ServerProfile getProfile(SmartyServer srv) {
        ServerProfile profile = new ServerProfile();
        profile.setName(srv.name);
        profile.setTitle(srv.name);
        profile.setVersion(srv.version);
        profile.setIp(srv.address);
        profile.setPort(srv.port);
        profile.setAssetDir(srv.name);
        profile.setExtraCheckSum(srv.extraCheckSum);
        profile.setOptionalModsData(srv.optionalMods);

        return profile;
    }
}

File: client-launcher/src/main/java/hivens/launcher/ServerListService.kt
package hivens.launcher

import hivens.core.api.IServerListService
import hivens.core.api.SmartyNetworkService
import hivens.core.api.dto.SmartyServer
import hivens.core.api.model.ServerProfile
import java.util.concurrent.CompletableFuture

// Внедряем зависимость через конструктор
class ServerListService(private val networkService: SmartyNetworkService) : IServerListService {

    // Кэш в памяти
    private var cachedProfiles: List<ServerProfile>? = null

    override fun fetchProfiles(): CompletableFuture<List<ServerProfile>> {
        // Если кэш есть - отдаем мгновенно
        if (!cachedProfiles.isNullOrEmpty()) {
            return CompletableFuture.completedFuture(cachedProfiles)
        }

        return CompletableFuture.supplyAsync {
            val profiles = ArrayList<ServerProfile>()
            try {
                // Вызываем метод, который мы написали в SmartyNetworkService
                val rawServers = networkService.getServers()
                for (srv in rawServers) {
                    profiles.add(getProfile(srv))
                }
                // Сохраняем в кэш
                this.cachedProfiles = profiles
            } catch (e: Exception) {
                System.err.println("Error fetching profiles: " + e.message)
            }
            profiles
        }
    }

    private fun getProfile(srv: SmartyServer): ServerProfile {
        return ServerProfile().apply {
            name = srv.name ?: "Unknown"
            title = srv.name
            version = srv.version ?: "1.7.10"
            // В DTO SmartyServer (Kotlin версия) поля называются address и port
            ip = srv.address ?: "localhost"
            port = srv.port
            assetDir = srv.name ?: "Unknown"
            extraCheckSum = srv.extraCheckSum
            optionalModsData = srv.optionalMods
        }
    }
}
